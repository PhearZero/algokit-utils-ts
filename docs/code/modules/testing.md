[@algorandfoundation/algokit-utils](../README.md) / testing

# Module: testing

## Table of contents

### Classes

- [TestLogger](../classes/testing.TestLogger.md)
- [TransactionLogger](../classes/testing.TransactionLogger.md)

### Functions

- [algoKitLogCaptureFixture](testing.md#algokitlogcapturefixture)
- [algorandFixture](testing.md#algorandfixture)
- [getTestAccount](testing.md#gettestaccount)
- [runWhenIndexerCaughtUp](testing.md#runwhenindexercaughtup)

## Functions

### algoKitLogCaptureFixture

▸ **algoKitLogCaptureFixture**(): [`AlgoKitLogCaptureFixture`](../interfaces/types_testing.AlgoKitLogCaptureFixture.md)

Creates a test fixture for capturing AlgoKit logs.

**`Example`**

```typescript
const logs = algoKitLogCaptureFixture()

beforeEach(logs.beforeEach)
afterEach(logs.afterEach)

test('My test', () => {
    const capturedLogs = logs.testLogger.capturedLogs
})
```

#### Returns

[`AlgoKitLogCaptureFixture`](../interfaces/types_testing.AlgoKitLogCaptureFixture.md)

The fixture

#### Defined in

[testing/fixtures/algokit-log-capture-fixture.ts:21](https://github.com/algorandfoundation/algokit-utils-ts/blob/main/src/testing/fixtures/algokit-log-capture-fixture.ts#L21)

___

### algorandFixture

▸ **algorandFixture**(`fixtureConfig?`): [`AlgorandFixture`](../interfaces/types_testing.AlgorandFixture.md)

Creates a test fixture for automated testing against Algorand.
By default it tests against a default LocalNet instance, but you can pass in an algod and indexer if you want to test against, say, TestNet.

**`Example`**

```typescript
const algorand = algorandFixture()

beforeEach(algorand.beforeEach)

test('My test', () => {
    const {algod, indexer, testAccount, ...} = algorand.context
})
```

#### Parameters

| Name | Type | Description |
| :------ | :------ | :------ |
| `fixtureConfig?` | [`AlgorandFixtureConfig`](../interfaces/types_testing.AlgorandFixtureConfig.md) | The fixture configuration |

#### Returns

[`AlgorandFixture`](../interfaces/types_testing.AlgorandFixture.md)

The fixture

#### Defined in

[testing/fixtures/algorand-fixture.ts:22](https://github.com/algorandfoundation/algokit-utils-ts/blob/main/src/testing/fixtures/algorand-fixture.ts#L22)

___

### getTestAccount

▸ **getTestAccount**(`param0`, `algod`, `kmd?`): `Promise`<`Account`\>

Creates an ephemeral Algorand account for the purposes of testing.
Returns a newly created random test account that is funded from the dispenser

**`See`**

 - DO NOT USE THIS TO CREATE A MAINNET ACCOUNT!
Note: By default this will log the mnemonic of the account.
 - 

#### Parameters

| Name | Type | Description |
| :------ | :------ | :------ |
| `param0` | [`GetTestAccountParams`](../interfaces/types_testing.GetTestAccountParams.md) | The config for the test account to generate |
| `algod` | `default` | An algod client |
| `kmd?` | `default` | A KMD client, if not specified then a default KMD client will be loaded from environment variables |

#### Returns

`Promise`<`Account`\>

The account, with private key loaded

#### Defined in

[testing/account.ts:17](https://github.com/algorandfoundation/algokit-utils-ts/blob/main/src/testing/account.ts#L17)

___

### runWhenIndexerCaughtUp

▸ **runWhenIndexerCaughtUp**<`T`\>(`run`): `Promise`<`T`\>

Runs the given indexer call until a 404 error is no longer returned.
Tried every 200ms up to 20 times.
Very rudimentary implementation designed for automated testing.

**`Example`**

```typescript
const transaction = await runWhenIndexerCaughtUp(() => indexer.lookupTransactionByID(txnId).do())
```

#### Type parameters

| Name |
| :------ |
| `T` |

#### Parameters

| Name | Type | Description |
| :------ | :------ | :------ |
| `run` | () => `Promise`<`T`\> | The code to run |

#### Returns

`Promise`<`T`\>

The result (as a promise), or throws if the indexer didn't catch up in time

#### Defined in

[testing/indexer.ts:11](https://github.com/algorandfoundation/algokit-utils-ts/blob/main/src/testing/indexer.ts#L11)
