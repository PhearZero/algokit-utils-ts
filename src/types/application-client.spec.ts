import { describe, test } from '@jest/globals'
import { getApplicationAddress } from 'algosdk'
import invariant from 'tiny-invariant'
import * as algokit from '../'
import { getBareCallContractData } from '../../tests/example-contracts/bare-call/contract'
import { algorandFixture } from '../testing'
import { AppSpec } from './appspec'

describe('application-client', () => {
  const localnet = algorandFixture()
  beforeEach(localnet.beforeEach, 10_000)

  let appSpec: AppSpec
  beforeAll(async () => {
    appSpec = (await getBareCallContractData()).appSpec
  })

  test('Create app', async () => {
    const { algod, indexer, testAccount } = localnet.context

    const client = algokit.getApplicationClient(
      {
        app: appSpec,
        sender: testAccount,
        creatorAddress: testAccount.addr,
        indexer,
      },
      algod,
    )

    const app = await client.create({
      //allowUpdate: true,
      //allowDelete: true,
      deployTimeParameters: {
        // It should strip off the TMPL_
        TMPL_UPDATABLE: 0,
        DELETABLE: 0,
        VALUE: 1,
      },
    })

    expect(app.appId).toBeGreaterThan(0)
    expect(app.appAddress).toBe(getApplicationAddress(app.appId))
    expect(app.confirmation?.['application-index']).toBe(app.appId)
  })

  test('Deploy app', async () => {
    const { algod, indexer, testAccount } = localnet.context

    const client = algokit.getApplicationClient(
      {
        app: appSpec,
        sender: testAccount,
        creatorAddress: testAccount.addr,
        indexer,
      },
      algod,
    )

    const app = await client.deploy({
      version: '1.0',
      deployTimeParameters: {
        VALUE: 1,
      },
    })

    invariant(app.operationPerformed === 'create')
    expect(app.appId).toBeGreaterThan(0)
    expect(app.appAddress).toBe(getApplicationAddress(app.appId))
    expect(app.confirmation?.['application-index']).toBe(app.appId)
  })

  test('Create then call app', async () => {
    const { algod, testAccount } = localnet.context
    const client = algokit.getApplicationClient(
      {
        app: appSpec,
        sender: testAccount,
        id: 0,
      },
      algod,
    )
    await client.create({
      deployTimeParameters: {
        UPDATABLE: 0,
        DELETABLE: 0,
        VALUE: 1,
      },
    })

    const call = await client.call({
      method: 'call',
      methodArgs: ['test'],
    })

    invariant(call.return)
    expect(call.return.decodeError).toBeUndefined()
    expect(call.return.returnValue).toBe('Hello, test')
  })
})
