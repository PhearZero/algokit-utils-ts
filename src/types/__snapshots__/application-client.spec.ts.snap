// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`application-client Errors Display nice error messages when there is a logic error 3`] = `
"INFO: Idempotently deploying app "TestingApp" from creator ACCOUNT_1 using 932 bytes of teal code and 4 bytes of teal code
INFO: App TestingApp not found in apps created by ACCOUNT_1; deploying app with version 1.0.
INFO: Sent transaction ID TXID_1 appl from ACCOUNT_1
DEBUG: Created app APP_1 from creator ACCOUNT_1
DEBUG: Received error executing Atomic Transaction Composer and debug flag enabled; attempting dry run to get more information
ERROR: Received the following application error when executing dry run of transaction | [{"trace":"pc# |ln# |source                            |scratch |stack\\n1   |1   |intcblock 0 1 10 5 1 1            |        |[]\\n9   |2   |bytecblock 0x 0x151f7c75          |        |[]\\n17  |3   |txn NumAppArgs                    |        |[]\\n19  |4   |intc_0 // 0                       |        |[1]\\n20  |5   |==                                |        |[1, 0]\\n21  |6   |bnz label1                        |        |[0]\\n24  |7   |txna ApplicationArgs 0            |        |[]\\n27  |8   |pushbytes 0xf17e80a5 // 0xf17e... |        |[0x44d0da0d]\\n33  |9   |==                                |        |[0x44d0da0d, 0xf17e80a5]\\n34  |10  |bnz label2                        |        |[0]\\n37  |11  |txna ApplicationArgs 0            |        |[]\\n40  |12  |pushbytes 0x0a92a81e // 0x0a92... |        |[0x44d0da0d]\\n46  |13  |==                                |        |[0x44d0da0d, 0x0a92a81e]\\n47  |14  |bnz label3                        |        |[0]\\n50  |15  |txna ApplicationArgs 0            |        |[]\\n53  |16  |pushbytes 0xa4cf8dea // 0xa4cf... |        |[0x44d0da0d]\\n59  |17  |==                                |        |[0x44d0da0d, 0xa4cf8dea]\\n60  |18  |bnz label4                        |        |[0]\\n63  |19  |txna ApplicationArgs 0            |        |[]\\n66  |20  |pushbytes 0xcec2834a // 0xcec2... |        |[0x44d0da0d]\\n72  |21  |==                                |        |[0x44d0da0d, 0xcec2834a]\\n73  |22  |bnz label5                        |        |[0]\\n76  |23  |txna ApplicationArgs 0            |        |[]\\n79  |24  |pushbytes 0xa4b4a230 // 0xa4b4... |        |[0x44d0da0d]\\n85  |25  |==                                |        |[0x44d0da0d, 0xa4b4a230]\\n86  |26  |bnz label6                        |        |[0]\\n89  |27  |txna ApplicationArgs 0            |        |[]\\n92  |28  |pushbytes 0x44d0da0d // 0x44d0... |        |[0x44d0da0d]\\n98  |29  |==                                |        |[0x44d0da0d, 0x44d0da0d]\\n99  |30  |bnz label7                        |        |[1]\\n246 |115 |txn OnCompletion                  |        |[]\\n248 |116 |intc_0 // 0                       |        |[0]\\n249 |117 |==                                |        |[0, 0]\\n250 |118 |txn ApplicationID                 |        |[1]\\n252 |119 |intc_0 // 0                       |        |[1, APP_1]\\n253 |120 |!=                                |        |[1, APP_1, 0]\\n254 |121 |&&                                |        |[1, 1]\\n255 |122 |assert                            |        |[1]\\n256 |123 |callsub label16                   |        |[]\\n779 |399 |proto 0 0                         |        |[]\\n782 |400 |intc_0 // 0                       |        |[]\\n783 |401 |assert                            |        |[0]\\n783 |401 |!! assert failed pc=783 !!        |        |[0]\\n"}]"
`;
